feat: Add new ptn modules `avm/ptn/azd/ml-project`.

Menghua1 opened the pull request:
## Description
Fixes https://github.com/Azure/Azure-Verified-Modules/issues/1228.
<!--
>Thank you for your contribution !
> Please include a summary of the change and which issue is fixed.
> Please also include the context.
> List any dependencies that are required for this change.

Fixes #123
Fixes #456
Closes #123
Closes #456
-->

## Pipeline Reference

<!-- Insert your Pipeline Status Badge below -->

| Pipeline |
| -------- |
|   [![avm.ptn.azd.ml-project](https://github.com/Menghua1/bicep-registry-modules/actions/workflows/avm.ptn.azd.ml-project.yml/badge.svg)](https://github.com/Menghua1/bicep-registry-modules/actions/workflows/avm.ptn.azd.ml-project.yml)     |

## Type of Change

<!-- Use the checkboxes [x] on the options that are relevant. -->

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

<!--  Please keep up to date with the contribution guide at https://aka.ms/avm/contribute/bicep -->

@jongio for notification.

Menghua1 commented:
@AlexanderSehr Could you help review this PR?

AlexanderSehr commented:
> @AlexanderSehr Could you help review this PR?

Hey @Menghua1,
as I'll be on vacation from this weekend for 2-ish weeks, it would probably make sense if another reviewer takes this and sees it through (cc: @Azure/avm-module-reviewers-bicep). Otherwise we may risk that I'll leave a comment and won't be able to resolve it before October 😄

Menghua1 commented:
> Hey @Menghua1, as I'll be on vacation from this weekend for 2-ish weeks, it would probably make sense if another reviewer takes this and sees it through (cc: @Azure/avm-module-reviewers-bicep). Otherwise we may risk that I'll leave a comment and won't be able to resolve it before October 😄

Hi AlexanderSehr,

Thank you for letting me know about your upcoming vacation. I understand the need to have another reviewer take over. Could you please help me ask if anyone from the @Azure/avm-module-reviewers-bicep group can review the PR promptly?

eriqua commented on .github/workflows/avm.ptn.azd.ml-hub-dependencies.yml:
Please remove as not part of this PR

eriqua commented on avm/ptn/azd/ml-project/main.bicep:
```diff
@@ -0,0 +1,171 @@
+  }
+}
+
+resource keyVault 'Microsoft.KeyVault/vaults@2022-07-01' existing = {
```
Probably cosmetic, but imho it would help the readability of the module if resource and module declaration followed the deployment order

eriqua commented on avm/ptn/azd/ml-project/main.bicep:
```diff
@@ -0,0 +1,171 @@
+
+@description('Optional. The role to assign. You can provide either the display name of the role definition, the role definition GUID, or its fully qualified ID in the following format: \'/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11\'.')
+param roleDefinitionIdOrName array = [
+  'f6c7c914-8db3-469d-8ca1-694a8f32e121'
```
Since these are default values, I'd suggest mentioning the user friendly corresponding name in the description or at least as a comment close to the Id

eriqua commented on avm/ptn/azd/ml-project/main.bicep:
```diff
@@ -0,0 +1,171 @@
+}
+
+@description('Required. The name of the user assigned identity.')
+param userAssignedtName string
```
Please correct typo in all instances

eriqua commented on avm/ptn/azd/ml-project/tests/e2e/defaults/dependencies.bicep:
```diff
@@ -0,0 +1,63 @@
+  }
+}
+
+@description('The name of the AI Studio Hub Resource.')
```
Is this output used? If not I'd suggest to remove it for cleaner code

eriqua commented on avm/ptn/azd/ml-project/tests/e2e/defaults/dependencies.bicep:
```diff
@@ -0,0 +1,63 @@
+  }
+}
+
+module storageAccount 'br/public:avm/res/storage/storage-account:0.9.1' = {
```
Also in this case it would help readability to order modules based on their order of deployment (e.g. put storage and kv before hub)

ReneHezser commented:
@Menghua1 
We have added examples for Bicep parameter files to the Readme. This has been applied to all published modules but needs to be done for PRs as well. Can you please run the Set-AVMModule utility as detailed [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/generate-bicep-module-files/). It is required for the validation pipeline to succeed and the contribution to be published.

Please reach out if any support is needed.

Menghua1 commented:
@eriqua I have updated the code based on your comments, please review it again, thank you!

Menghua1 commented:
> @Menghua1 We have added examples for Bicep parameter files to the Readme. This has been applied to all published modules but needs to be done for PRs as well. Can you please update your branch and run the Set-AVMModule utility as detailed [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/generate-bicep-module-files/). It is required for the validation pipeline to succeed and the contribution to be published.
> 
> Please reach out if any support is needed.

@ReneHezser Thank you for your guidance. I have updated branch and re-ran the Set-AVMModule utility.

eriqua commented on .github/workflows/avm.ptn.azd.ml-ai-environment.yml:
This workflow file should be removed since unrelated to this PR

eriqua commented on avm/ptn/azd/ml-project/main.bicep:
```diff
@@ -0,0 +1,171 @@
+metadata name = 'Machine Learning workspace'
```
Ca we please specify Azd in the readme title to avoid any confusion with the resource module?

eriqua commented on avm/ptn/azd/ml-project/main.bicep:
```diff
@@ -0,0 +1,171 @@
+  'Enabled'
+  'Disabled'
+])
+param publicNetworkAccess string = 'Enabled'
```
Is this intended to be enabled by default?

Menghua1 replied:
Yes, it is aligned with the default values ​​of the [project.bicep](https://github.com/Azure/azure-dev/blob/main/templates/common/infra/bicep/core/ai/project.bicep#L17) file parameters in infra/core.

eriqua commented on avm/ptn/azd/ml-project/main.json:
```diff
@@ -0,0 +1,7427 @@
+  "metadata": {
+    "_generator": {
+      "name": "bicep",
+      "version": "0.28.1.47646",
```
Not mandatory, just a heads up that a newer version of bicep is available, in case you'd like to upgrade it

eriqua commented:
> @eriqua I have updated the code based on your comments, please review it again, thank you!

Just a couple of additional comment @Menghua1, thank you for the great work

AlexanderSehr commented on avm/ptn/azd/ml-project/main.bicep:
```diff
@@ -0,0 +1,171 @@
+
+module project 'br/public:avm/res/machine-learning-services/workspace:0.7.0' = {
+  name: '${uniqueString(deployment().name, location)}-project'
+  params: {
```
Misses passthru of `enableTelemetry` parameter

AlexanderSehr replied:
@Menghua1, please note, you did not add a passthru (`enableTelemetry: enableTelemetry`) but instead added the telemetry as a condition for the module's deployment (which you certainly don't want) 😉 

Please fix this for all module references. Also, I'd recommend to not resolve comments yourself, but simply comment them once you tackled them. Otherwise re-review are quite painful.

AlexanderSehr commented on avm/ptn/azd/ml-project/main.bicep:
```diff
@@ -0,0 +1,171 @@
+  name: '${uniqueString(deployment().name, location)}-keyvault'
+  params: {
+    name: keyVault.name
+    accessPolicies: [
```
Misses passthru of `enableTelemetry` parameter

AlexanderSehr replied:
Not fixed (see other comment)

AlexanderSehr commented on avm/ptn/azd/ml-project/main.bicep:
```diff
@@ -0,0 +1,171 @@
+
+module mlServiceRoleAssigned 'br/public:avm/res/managed-identity/user-assigned-identity:0.4.0' = {
+  name: '${uniqueString(deployment().name, location)}-roleassignment'
+  params: {
```
Misses passthru of `enableTelemetry` parameter

AlexanderSehr replied:
Not fixed (see other comment)

Menghua1 commented:
@eriqua and @AlexanderSehr  Thank you for your feedback! I have updated the code based on your comments, please review it again, thank you!

Menghua1 commented:
@AlexanderSehr I apologize for the oversight. Thank you for your feedback! I have submitted a new commit to address the issues correctly. Thank you very much.

AlexanderSehr commented:
> @AlexanderSehr I apologize for the oversight. Thank you for your feedback! I have submitted a new commit to address the issues correctly. Please review it.

Much appreciated, thank you @Menghua1 . @eriqua will see the PR through.
One last note: Please resolve the conflicts 😉

Menghua1 commented:
> Much appreciated, thank you @Menghua1 . @eriqua will see the PR through. One last note: Please resolve the conflicts the merge of the `azd/ml-hub-dependencies` PR introduced to this PR 😉

OK, thank you for your guidance. @eriqua the conflict has been resolved. Please review it, thank you!😉
